vimPlugins: rename nvimSkipModule into nvimSkipModules

teto opened the pull request:
It's confusing to have a polymorphic behavior for nvimSkipModule and it is an implementation detail. Let's not force ourselves to keep supporting it. 
Support and showcase just the list format to avoid confusion.The naming for lists is usually plural (lib.optional vs lib.optionals for instnace)


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

teto committed:
vimPlugins: rename nvimSkipModule into nvimSkipModules

It's confusing to have a polymorphic behavior for nvimSkipModule and it
is an implementation detail.
Let's not force ourselves to keep supporting both types and just support
lists, which imply plural in nixpkgs (e.g. "lib.optional" vs "lib.optionals").

Support and showcase just the list format to avoid confusion.

khaneliman commented:
This will silently break anyone using ‘buildVimPlugin’ outside Nixpkgs. We should deprecate the old name, not just drop it without notice.

teto commented:
damnn it seems already used in the wild https://github.com/search?q=nvimSkipModule&type=code .
How do you suggest to fix that ? would
```
if [ -v $nvimSkipModule ]; then
    
      nvimSkipModules=nvimSkipModule
      nix_warn "nvimSkipModule got renamed to nvimSkipModules"
fi
```
be ok ?

mrcjkb commented:
If it isn't in one of the stable channels, I wouldn't worry too much about breaking compatibility.
After all, the name `-unstable` implies that things might break.

teto committed:
vimPlugins: add warning for nvimSkipModule deprecation

It is best effort since this was not part of a release yet.
I tried to do display the warning at eval time but it ended up being harder than expected.
I tried to rewrite buildVimPlugin with lib.extendMkDerivation to access the fixed point to no avail.
I think it is something worth doing but in a separate PR.

Here is a quickfix to avoid it ending up in release.

khaneliman commented:
> it would have been cleverer to rename it at eval time to avoid a rebuild. I might change it later, anyone who feels like doing it feel free to push over the PR ^^''

I wouldn't worry about rebuilds, most of our `vimPlugins` aren't built on Hydra atm

teto commented:
it's more convoluted than I expected to do this nix side. I couldn't have the eval see nvimSkipModule.
I tried to rewrite buildVimPlugin with lib.extendMkDerivation to access the fixed point to no avail. I think it is something worth doing but in a separate PR.

I could go full nuke and `exit 1` in the if but let's assume the build goes on: 
case1: build works. Ok so nvimSkipModule was not needed
case2: build fails: person looks at the logs and sees the message. We are still ok.

khaneliman commented on doc/languages-frameworks/neovim.section.md:
```diff
@@ -170,13 +170,12 @@ To only check a specific module, add it manually to the plugin definition [overr
-We can skip specific modules using `nvimSkipModule`. Similar to `nvimRequireCheck`, it accepts a single string or a list of strings.
-- `nvimSkipModule = MODULE;`
-- `nvimSkipModule = [ MODULE1 MODULE2 ];`
+We can skip specific modules using `nvimSkipModules`. Similar to `nvimRequireCheck`, it accepts a single string or a list of strings.
```
```suggestion
We can skip specific modules using `nvimSkipModules`. Similar to `nvimRequireCheck`, it accepts a list of strings.
```

khaneliman reviewed (approved):
LGTM, if we have time we can try and make the deprecation louder. But, at least this gets the property and docs updated before release

teto commented:
nixpkgs-review seems ok so far (built 1k plugins). Ty for reviews. I hope we can convert the check at build time by allowing `buildVimPlugin` to accept the finalAttrs pattern but until someone can do it, let's merge this to make sure the interface is correct.

llakala commented:
>  I hope we can convert the check at build time by allowing buildVimPlugin to accept the finalAttrs pattern but until someone can do it, let's merge this to make sure the interface is correct.

Want me to make an issue for this?

teto commented:
sure. Even better a PR ;)
